ui-trackoccupancydiagram/src/components/helpers/drawElements/drawOccupancyZonesTexts.ts

```diff
@@ -1,105 +1,151 @@
-import {
-  OCCUPANCY_ZONE_START,
-  MINUTES_TEXT_OFFSET,
-  STATION_TEXT_OFFSET,
-  FONTS,
-  COLORS,
-} from '../../consts';
+import { MINUTES_TEXT_OFFSET, STATION_TEXT_OFFSET, FONTS, COLORS } from '../../consts';
 import { drawText } from '../../utils';
 
 const BREAKPOINTS = {
   medium: 24,
   small: 4,
 };
 const STROKE_WIDTH = 4;
+const X_BACKGROUND_PADDING = 4;
 const X_INITIAL_POSITION_OFFSET = 8;
+const X_MEDIUM_POSITION_OFFSET_BACKGROUND = 20;
 const Y_INITIAL_POSITION_OFFSET = 5;
+const Y_INITIAL_POSITION_OFFSET_BACKGROUND = 18;
+const X_TROUGHTRAIN_PADDING = 2;
+const X_SELECTED_TROUGHTRAIN_PADDING = 8;
+const X_THROUGHTRAIN_OFFSET = 4;
+const Y_MEDIUM_POSITION_OFFSET_BACKGROUND = 32;
 const Y_MEDIUM_POSITION_OFFSET = 14;
+const ROTATE_VALUE = (-30 * Math.PI) / 180;
 
 const { SANS, MONO } = FONTS;
-const { WHITE_100, GREY_50, GREY_60, GREY_80 } = COLORS;
+const { WHITE_100, GREY_50, GREY_60, GREY_80, SELECTION_20 } = COLORS;
 
 export const drawOccupancyZonesTexts = ({
   ctx,
   zone,
-  arrivalTime,
-  departureTime,
+  arrivalTimePixel,
+  departureTimePixel,
+  yPosition,
+  isThroughTrain,
+  selectedTrainId,
 }: {
   ctx: CanvasRenderingContext2D;
   zone: {
+    id: string;
     arrivalTrainName: string;
     arrivalTime: Date;
     departureTime: Date;
     originStation?: string;
     destinationStation?: string;
   };
-  arrivalTime: number;
-  departureTime: number;
+  arrivalTimePixel: number;
+  departureTimePixel: number;
+  yPosition: number;
+  isThroughTrain: boolean;
+  selectedTrainId: string;
 }) => {
-  const zoneOccupancyLength = departureTime - arrivalTime - STROKE_WIDTH;
+  const zoneOccupancyLength = departureTimePixel - arrivalTimePixel - STROKE_WIDTH;
 
   const isBelowBreakpoint = (breakpoint: keyof typeof BREAKPOINTS) =>
     zoneOccupancyLength < BREAKPOINTS[breakpoint];
 
-  const textLength = ctx.measureText(zone.originStation!).width;
-  const { xName, yName } = {
-    xName: isBelowBreakpoint('medium')
-      ? arrivalTime - textLength + STROKE_WIDTH
-      : arrivalTime + X_INITIAL_POSITION_OFFSET,
-    yName: isBelowBreakpoint('medium')
-      ? OCCUPANCY_ZONE_START - Y_MEDIUM_POSITION_OFFSET
-      : OCCUPANCY_ZONE_START - Y_INITIAL_POSITION_OFFSET,
+  const originTextLength = ctx.measureText(zone.originStation!).width;
+  const nameTextLength = ctx.measureText(zone.arrivalTrainName).width;
+
+  const { xOriginName, yOriginName } = {
```

Comment:
I lost almost 15min because I thought it was used to place the origin station name 🫠
```suggestion
  const { xArrivalTrainName, yArrivalTrainName } = {
```

Reply:
done